model/src/com/redhat/ceylon/model/typechecker/util/TypePrinter.java

```diff
@@ -483,13 +483,33 @@ private static boolean isTupleTypeWellformed(Type args) {
         if (elemtypes==null) {
             return false;
         }
-        Type t = union(elemtypes, unit);
-        Type typeArg = args.getTypeArgumentList().get(0);
-        if (typeArg==null) {
-            return false;
-        }
-        else {
-            return t.isExactly(typeArg);
+
+        int index = -1;
+        while (true) {
+            index++;
+            if (args.isTuple()) {
+                List<Type> tal = args.getTypeArgumentList();
+                if (tal.size() < 3) {
+                    return false;
+                }
+                Type t = union(elemtypes.subList(
+                        index, elemtypes.size()), unit);
+                Type typeArg = tal.get(0);
+                if (typeArg==null ||
+                        !t.isExactly(typeArg)) {
+                    return false;
+                }
+                // check Rest
+                args = tal.get(2);
+            }
+            else if (args.isEmpty()
+                    || args.isSequential()
+                    || args.isSequence()) {
+                return true;
```

Comment:
In theory, `args` will always be a `Tuple` in the first iteration of this loop, which was the only iteration the previous code bothered with. And, if it's exactly one of those three, it can always be abbreviated.

The "in theory" part is that callers pre-check, and also line 469 makes sure there are at least 3 type args.